torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp

```diff
@@ -0,0 +1,637 @@
+    max_input_dtype_size = std::max(
+        max_input_dtype_size,
+        (int64_t)dataTypeSize(inp->getDataType().value(), index_type));
+    n_input_tensors++;
+  }
+
+  auto ref_root = largest1->getMaybeRFactorDomain();
+  int64_t n_elems = 1;
+  for (size_t ref_i = 0; ref_i < ref_root.size(); ref_i++) {
+    auto inferred_val =
+        runtime_info.expressionEvaluator().evaluate(ref_root[ref_i]->extent());
+    TORCH_INTERNAL_ASSERT(
+        inferred_val.has_value(),
+        "Error inferring size for pointwise scheduler: ",
+        ref_root[ref_i]->extent()->toInlineString());
+    n_elems *= inferred_val.value();
+  }
+
+  TransposeParams params;
+  params.tag = "Transpose heuristics";
+
+  // Note [vectorization and unroll of input and output]
+  //
+  // The choice of vectorization size, block size and tile sizes needs to be
+  // consistent with each other. Consider the following:
+  //
+  // The number of threads in one block is
+  //   num_threads = blockDim.x * blockDim.y
+  // and the number of elements per each tile is
+  //   num_elems_per_tile = params.tile_size1 * params.tile_size2
+  // So each thread needs to process
+  //   num_elems_per_thread = num_elems_per_tile / num_threads
+  // elements. That is, once the tile sizes and block size are determined, the
+  // `num_elems_per_thread` is determined, regardless of vectorizability of
+  // input/output tensors.
+  //
+  // To make the selection of tile sizes othogonal to vectorizability, we
+  // support having both vectorization and unrolling in the same tensor. For
+  // example, if we have num_elems_per_tile == 1024 and num_threads = 256, then
+  // we have num_elems_per_thread being 4. And if we have vector size 2, then we
+  // will do unroll 2 * vectorize 2 at the same tensor.
+  //
+  // Also, since the inner most dim of different groups are not the same, it is
+  // natural to consider their vectorizability separately and allow them to have
+  // different vectorize/unroll sizes.
+
+  constexpr int64_t kSixteen = 16; // clang tidy
+
+  auto max_unroll_factor = ceilDiv(
+      // Available unrolling based on size of data type
+      (int64_t)kSixteen / max_input_dtype_size,
+      // Reduce max unrolling factor if we have many inputs/outputs to unroll
+      // as it could start consuming a lot of registers.
+      std::max(
+          (scheduler_utils::lastPow2(
+               (int64_t)grouped_inputs_outputs[0].size() +
+               (int64_t)grouped_inputs_outputs[1].size()) >>
+           2),
+          (int64_t)1));
+
+  // Don't unroll at the cost of getting a full wave on the GPU
+  auto max_unroll_factor_occupancy = ceilDiv(
+      n_elems,
+      device_multiprocessor_count * params.tile_size1 * params.tile_size2);
+  max_unroll_factor = std::min(max_unroll_factor, max_unroll_factor_occupancy);
+
+  // If we use RNG don't unroll so we can do correctness testing
+  if (fusion->isStochastic() && isDisabled(DisableOption::UnrollWithRng)) {
```

Comment:
I don't think we can validate transpose with RNG as it would depend on if the RNG is before or after the thread binding switch. This is the same for normalizations or reductions since we don't parallelize the same way we can't guarantee RNG consistency with eager mode.

Reply:
fixed